File: src/core/content/cyclesMossy.ts
/**
 * MOSSY MAYHEM'S LADDER - eight creatures, sixteen minutes, its own numbers.
 *
 * ---------------------------------------------------------------------------------------------
 * SEEDED FROM THE SCRAPYARD'S CURVE, AND NOW COMPLETELY INDEPENDENT OF IT
 * ---------------------------------------------------------------------------------------------
 * The HP, speed, damage and XP columns started as copies of the Scrapyard's, because that curve is
 * measured and this one is not, and shipping a second level on unproven pacing would make every
 * later "is Mossy too hard" question unanswerable. They are COPIES AND NOT REFERENCES, on purpose
 * and by instruction: retuning cycle 6 here must not be able to touch the yard, and the way to
 * guarantee that is for there to be no shared symbol to retune.
 *
 * So the two ladders will drift, and that is the intended end state rather than an accident to be
 * cleaned up later. When Mossy's numbers move, move them here and nowhere else.
 *
 * ---------------------------------------------------------------------------------------------
 * WHAT MAKES A MOSSY CYCLE DIFFERENT FROM A SCRAPYARD ONE
 * ---------------------------------------------------------------------------------------------
 * A Scrapyard cycle names a HULL and a TIER and derives its three ranks by recolouring one atlas
 * frame. There is nothing to recolour here, so a Mossy cycle names its three creatures OUTRIGHT -
 * `regular`, `elite`, `boss`, three ids into `MOSS_CREATURES`. Three consequences worth knowing:
 *
 *   - A cycle may repeat one creature across all three ranks, and three of them do. That is the
 *     "one creature, three sizes" answer, and it costs a line rather than a mechanism.
 *   - A cycle's ranks may be unrelated art if that ever reads better. Nothing enforces a family.
 *   - THE BODY CLASS IS AUTHORED, not derived. The Scrapyard reads it off measured sprite area
 *     because its atlas has a hull column to read; DCSS tiles are all 32x32 with no such structure,
 *     so a pixel-area rule would call a hydra and a jackal the same size. Pick the class that
 *     matches the drawSize in `creaturesMossy.ts` - the collision radius comes from the class and
 *     the sprite comes from the creature, and those two disagreeing is the one bug here that
 *     players notice immediately.
 *
 * ---------------------------------------------------------------------------------------------
 * THE SHAPE OF THE EIGHT
 * ---------------------------------------------------------------------------------------------
 * Same principle the Scrapyard's ladder is built on and worth restating, because it is the reason
 * the columns below are not all monotone: only HP and XP climb every rung. Speed and contact
 * damage move up AND down, so each cycle asks a different question instead of asking the last
 * one's louder.
 *
 *      1 Sporeling    Slow, soft, and there are a lot of them. One enemy for two minutes.
 *      2 Swarm        FAST AND FLIMSY. It teaches that distance is not safety.
 *      3 Formless     SLOW AND FAT. You can walk away from these; you cannot ignore them.
 *      4 Pack         THE FASTEST THING ON THE MAP, and lighter than the Formless before it.
 *      5 Vine Stalker Slams the brakes and nearly doubles the bite. First one that hurts to touch.
 *      6 Draconian    Quick AGAIN, and hits hardest of anything so far.
 *      7 Golem        Tanky and unhurried.
 *      8 Wyrm         A WALL. Slowest and by far the heaviest.
 *
 * Invariant K holds with room: the fastest thing here is a `swift` Pack regular at 71 x 1.18 x the
 * speed ramp = 100.6 u/s, against a 195 u/s mech. tests/levels.test.ts checks it per level rather
 * than trusting that one measurement covers both.
 */

import {
  RANK_BOSS,
  RANK_ELITE,
  RANK_REGULAR,
  type ResolvedCycle,
} from './cycles.js';
import { MOSS } from './creaturesMossy.js';
import { ARCH_BRUISER, ARCH_GRUNT, ARCH_RUNT, type Archetype } from './enemyCatalog.js';

export interface MossyCycleDef {
  readonly name: string;
  /**
   * Body class: collision radius, mass and contact interval. AUTHORED, not derived - see the
   * header. Must agree with the creature's `drawSize` in creaturesMossy.ts.
   */
  readonly archetype: Archetype;
  /** Creature ids, one per rank. May repeat - three of these cycles use one creature throughout. */
  readonly regular: number;
  readonly elite: number;
  readonly boss: number;
  /** Regular HP at cycle START, before the within-cycle ramp. */
  readonly hp: number;
  readonly speed: number;
  readonly contactDamage: number;
  readonly xp: number;
  /** P(a regular rolls a non-plain flavour). Zero in cycle 0: the first minute is ONE enemy. */
  readonly variantChance: number;
}

export const MOSS_LADDER: readonly MossyCycleDef[] = Object.freeze([
  Object.freeze({
    name: 'Sporeling', archetype: ARCH_RUNT as Archetype,
    regular: MOSS.SPORELING, elite: MOSS.SPORELING, boss: MOSS.SPORELING,
    hp: 22, speed: 56, contactDamage: 5, xp: 1, variantChance: 0,
  }),
  Object.freeze({
    name: 'Swarm', archetype: ARCH_RUNT as Archetype,
    regular: MOSS.BLOWFLY, elite: MOSS.KILLER_BEE, boss: MOSS.MOSQUITO,
    hp: 34, speed: 68, contactDamage: 6, xp: 2, variantChance: 0.1,
  }),
  Object.freeze({
    name: 'Formless', archetype: ARCH_GRUNT as Archetype,
    regular: MOSS.JELLY, elite: MOSS.OOZE, boss: MOSS.SHELLBACK,
    hp: 56, speed: 54, contactDamage: 9, xp: 3, variantChance: 0.16,
  }),
  Object.freeze({
    name: 'Pack', archetype: ARCH_RUNT as Archetype,
    regular: MOSS.JACKAL, elite: MOSS.RAIJU, boss: MOSS.HELLHOUND,
    hp: 66, speed: 71, contactDamage: 8, xp: 4, variantChance: 0.22,
  }),
  Object.freeze({
    name: 'Vine Stalker', archetype: ARCH_GRUNT as Archetype,
    regular: MOSS.VINE_STALKER, elite: MOSS.VINE_STALKER, boss: MOSS.VINE_STALKER,
    hp: 104, speed: 53, contactDamage: 14, xp: 6, variantChance: 0.26,
  }),
  Object.freeze({
    name: 'Draconian', archetype: ARCH_BRUISER as Archetype,
    regular: MOSS.DRACONIAN, elite: MOSS.DRACONIAN, boss: MOSS.DRACONIAN,
    hp: 118, speed: 65, contactDamage: 18, xp: 8, variantChance: 0.3,
  }),
  Object.freeze({
    name: 'Golem', archetype: ARCH_GRUNT as Archetype,
    regular: MOSS.EARTH_ELEMENTAL, elite: MOSS.STONE_GOLEM, boss: MOSS.IRON_GOLEM,
    hp: 172, speed: 57, contactDamage: 15, xp: 11, variantChance: 0.32,
  }),
  Object.freeze({
    name: 'Wyrm', archetype: ARCH_BRUISER as Archetype,
    regular: MOSS.DRAGON, elite: MOSS.GOLDEN_DRAGON, boss: MOSS.HYDRA,
    hp: 225, speed: 50, contactDamage: 22, xp: 15, variantChance: 0.34,
  }),
] as const) as readonly MossyCycleDef[];

/**
 * Per-cycle multipliers applied past the end of the authored ladder. Mossy's own copies: a run
 * longer than sixteen minutes should be able to get harder here at a different rate than it does
 * in the yard, without anybody having to notice that a shared constant existed.
 */
const EXTRA_HP_MUL = 1.45;
const EXTRA_XP_MUL = 1.4;
const EXTRA_DMG_MUL = 1.2;

/**
 * Fills `out` with cycle `index`'s creature.
 *
 * Past the authored ladder it repeats the LAST cycle with compounding multipliers. It does not
 * rotate anything: the Scrapyard rotates its recolour there so extrapolated cycles keep looking
 * new, and that is only possible because it has four paints of every hull. The honest answer here
 * is that minute 17 is another Wyrm, harder - and dressing that up would mean pointing cycle 9 at
 * a creature the ladder never designed a place for.
 *
 * `Math.pow` is banned in core (implementation-defined, so V8 and JSC may disagree in the last ulp
 * and one ulp of enemy HP is a divergent replay), so the compounding is a loop of exactly-rounded
 * multiplies - which runs at most once per 120 seconds.
 */
export function resolveMossyCycle(index: number, out: ResolvedCycle): void {
  const n = MOSS_LADDER.length;
  const i = index < n ? index : n - 1;
  const extra = index < n ? 0 : index - (n - 1);
  const def = MOSS_LADDER[i];

  let hp = def.hp;
  let xp = def.xp;
  let dmg = def.contactDamage;
  for (let k = 0; k < extra; k++) {
    hp *= EXTRA_HP_MUL;
    xp *= EXTRA_XP_MUL;
    dmg *= EXTRA_DMG_MUL;
  }

  out.index = index;
  out.name = def.name;
  out.archetype = def.archetype;
  out.hp = hp;
  out.speed = def.speed;
  out.contactDamage = dmg;
  out.xp = xp;
  out.variantChance = def.variantChance;

  out.typeByRank[RANK_REGULAR] = def.regular;
  out.typeByRank[RANK_ELITE] = def.elite;
  out.typeByRank[RANK_BOSS] = def.boss;
}

File: cs/src/Scrapyard.Core/CyclesMossy.cs
namespace Scrapyard.Core;

/// <summary>
/// Creature ids for Mossy Mayhem. Port of <c>MOSS</c> in <c>content/creaturesMossy.ts</c>.
/// </summary>
/// <remarks>
/// Positional and arbitrary in themselves - id 3 here means nothing the Scrapyard's id 3 does, and
/// neither table can be renumbered by editing the other - but the NUMBER is real simulation state
/// once it lands in <c>EnemyPool.TypeId</c> and the hash. Only the ids are ported; the sprite frame
/// strings and draw sizes they name are presentation with no reader in <c>stepWorld</c> (the
/// archetype already supplies the collision radius - see the remarks on <see cref="WeaponDef"/> for
/// the general rule this follows).
/// </remarks>
public static class MossCreatures
{
    public const int Sporeling = 0;
    public const int Blowfly = 1;
    public const int KillerBee = 2;
    public const int Mosquito = 3;
    public const int Jelly = 4;
    public const int Ooze = 5;
    public const int Shellback = 6;
    public const int Jackal = 7;
    public const int Raiju = 8;
    public const int Hellhound = 9;
    public const int VineStalker = 10;
    public const int Draconian = 11;
    public const int EarthElemental = 12;
    public const int StoneGolem = 13;
    public const int IronGolem = 14;
    public const int Dragon = 15;
    public const int GoldenDragon = 16;
    public const int Hydra = 17;
}

/// <summary>
/// MOSSY MAYHEM'S LADDER. Port of <c>cyclesMossy.ts</c>.
/// </summary>
/// <remarks>
/// <para>
/// COPIED FROM THE SCRAPYARD'S CURVE AND NOW INDEPENDENT OF IT, deliberately: the HP/speed/damage/
/// xp columns started as copies because that curve is measured and this one is not, but they are
/// copies and not references so that retuning one can never touch the other.
/// </para>
/// <para>
/// The one structural difference from <see cref="ScrapyardLadder"/>: no hull/tier recolour, so
/// each rung names three creature ids OUTRIGHT (which may repeat - three of these cycles use one
/// creature at all three ranks), and past the authored table it repeats the LAST rung rather than
/// rotating a paint - there is no second colour to rotate to.
/// </para>
/// </remarks>
public static class MossyLadder
{
    public readonly struct Rung
    {
        public required string Name { get; init; }
        public required int Archetype { get; init; }
        public required int Regular { get; init; }
        public required int Elite { get; init; }
        public required int Boss { get; init; }
        public required double Hp { get; init; }
        public required double Speed { get; init; }
        public required double ContactDamage { get; init; }
        public required double Xp { get; init; }
        public required double VariantChance { get; init; }
    }

    public static readonly Rung[] All =
    {
        new() { Name = "Sporeling", Archetype = Archetypes.Runt, Regular = MossCreatures.Sporeling, Elite = MossCreatures.Sporeling, Boss = MossCreatures.Sporeling, Hp = 22, Speed = 56, ContactDamage = 5, Xp = 1, VariantChance = 0 },
        new() { Name = "Swarm", Archetype = Archetypes.Runt, Regular = MossCreatures.Blowfly, Elite = MossCreatures.KillerBee, Boss = MossCreatures.Mosquito, Hp = 34, Speed = 68, ContactDamage = 6, Xp = 2, VariantChance = 0.1 },
        new() { Name = "Formless", Archetype = Archetypes.Grunt, Regular = MossCreatures.Jelly, Elite = MossCreatures.Ooze, Boss = MossCreatures.Shellback, Hp = 56, Speed = 54, ContactDamage = 9, Xp = 3, VariantChance = 0.16 },
        new() { Name = "Pack", Archetype = Archetypes.Runt, Regular = MossCreatures.Jackal, Elite = MossCreatures.Raiju, Boss = MossCreatures.Hellhound, Hp = 66, Speed = 71, ContactDamage = 8, Xp = 4, VariantChance = 0.22 },
        new() { Name = "Vine Stalker", Archetype = Archetypes.Grunt, Regular = MossCreatures.VineStalker, Elite = MossCreatures.VineStalker, Boss = MossCreatures.VineStalker, Hp = 104, Speed = 53, ContactDamage = 14, Xp = 6, VariantChance = 0.26 },
        new() { Name = "Draconian", Archetype = Archetypes.Bruiser, Regular = MossCreatures.Draconian, Elite = MossCreatures.Draconian, Boss = MossCreatures.Draconian, Hp = 118, Speed = 65, ContactDamage = 18, Xp = 8, VariantChance = 0.3 },
        new() { Name = "Golem", Archetype = Archetypes.Grunt, Regular = MossCreatures.EarthElemental, Elite = MossCreatures.StoneGolem, Boss = MossCreatures.IronGolem, Hp = 172, Speed = 57, ContactDamage = 15, Xp = 11, VariantChance = 0.32 },
        new() { Name = "Wyrm", Archetype = Archetypes.Bruiser, Regular = MossCreatures.Dragon, Elite = MossCreatures.GoldenDragon, Boss = MossCreatures.Hydra, Hp = 225, Speed = 50, ContactDamage = 22, Xp = 15, VariantChance = 0.34 },
    };

    /// <summary>Mossy's own copies - retuning one ladder's extrapolation must not touch the other's.</summary>
    private const double ExtraHpMul = 1.45;
    private const double ExtraXpMul = 1.4;
    private const double ExtraDmgMul = 1.2;

    public static void Resolve(int index, ResolvedCycle outc)
    {
        int n = All.Length;
        int i = index < n ? index : n - 1;
        int extra = index < n ? 0 : index - (n - 1);
        ref readonly var def = ref All[i];

        double hp = def.Hp;
        double xp = def.Xp;
        double dmg = def.ContactDamage;
        for (int k = 0; k < extra; k++)
        {
            hp *= ExtraHpMul;
            xp *= ExtraXpMul;
            dmg *= ExtraDmgMul;
        }

        outc.Index = index;
        outc.Name = def.Name;
        outc.Archetype = def.Archetype;
        outc.Hp = hp;
        outc.Speed = def.Speed;
        outc.ContactDamage = dmg;
        outc.Xp = xp;
        outc.VariantChance = def.VariantChance;

        outc.TypeByRank[Ranks.Regular] = def.Regular;
        outc.TypeByRank[Ranks.Elite] = def.Elite;
        outc.TypeByRank[Ranks.Boss] = def.Boss;
    }
}

/// <summary>Mossy Mayhem.</summary>
public sealed class MossyMayhemLevel : ILevel
{
    public int CycleCount => MossyLadder.All.Length;

    /// <summary>
    /// A FLOCK OF FOUR - the only level in the game with one, because it is the only one whose
    /// terrain gives nothing back when it is broken. See <see cref="Sheep"/>.
    /// </summary>
    public int Sheep => 4;

    public void ResolveCycle(int index, ResolvedCycle outc) => MossyLadder.Resolve(index, outc);
}
